core/src/main/java/org/apache/spark/shuffle/sort/UnsafeShuffleWriter.java

```diff
@@ -310,7 +323,9 @@ void forceSorterToSpill() throws IOException {
         // strategies use different IO techniques.  We count IO during merge towards the shuffle
         // shuffle write time, which appears to be consistent with the "not bypassing merge-sort"
         // branch in ExternalSorter.
-        if (fastMergeEnabled && fastMergeIsSupported) {
+        if (pluggableWriteSupport != null) {
+          partitionLengths = mergeSpillsWithPluggableWriter(spills, compressionCodec);
```

Comment:
I think you also want to support the "fast" paths here.  I mentioned the transferTo part above, but even when working with output streams, you want to avoid decompressing and recompressing if possible, so you can just copy the bytes directly from one input stream to the other output stream -- thats the "Using fileStream-based fast merge" path.

(If it would be helpful, we can chat about this direclty)

Reply:
Compression is tricky, as is encryption really. The question is, should the underlying implementation of the API be the part that decides how to compress and encrypt the output in the remote storage? It's not immediately obvious. On one hand we can save work by taking the compressed and encrypted local files and passing them directly without unmarshalling them. On the other hand, should the implementation layer be allowed to decide to use its own encryption algorithms and compression codecs? That seems like a reasonable use case as well. But if the storage implementation is doing its own encryption and compression, isn't it encrypting and compressing already-encrypted and already-compressed data?

My intuition is that we want the implementation of the API to decide on its own encryption and compression, for example if it wants to do its own splitting of partitions and thus would prefer to encrypt and compress each block individually. In that case I don't think it makes a lot of sense to pass along encrypted and compressed bytes to this OutputStream. But again I think this is open to discussion.

Reply:
Do some of these questions around encryption and compression also influence how / if we support `transferTo` and channel-based writes? This part of the code is still something I'm trying to understand better.

Reply:
I don't think this really influences `transferTo` -- I think you can only use transferTo with the "fast path" here, but this is a good improvement even for the outputstream version.

I see your point about the external service maybe wanting to manage its own encryption and compression -- but I also don't really think you want to tackle that here.  I think it'll keep this first step much easier if we still leave that up to the spark shuffle writers.